libnm-core/nm-utils.c

```diff
@@ -6223,6 +6223,33 @@ nm_utils_format_variant_attributes (GHashTable *attributes,
 
 /*****************************************************************************/
 
+/*
+ * nm_utils_get_timestamp_msec():
+ *
+ * Gets current time in milliseconds of CLOCK_BOOTTIME.
+ *
+ * Returns: time in milliseconds
+ *
+ * Since: 1.12
+ */
+gint64
+nm_utils_get_timestamp_msec (void)
+{
+	struct timespec ts;
+
+	if (clock_gettime (CLOCK_BOOTTIME, &ts) == -1) {
+		/* The fallback to CLOCK_MONOTONIC is taken only if we're running on a
+		 * criminally old kernel, prior to 2.6.39 (released on 18 May, 2011).
+		 * That happens during buildcheck on old builders, we don't expect to
+		 * be actually runs on kernels that old. */
+		return g_get_monotonic_time () / 1000;
```

Comment:
should you not check `errno == EINVAL` and only perform the fallback then?
And on all other errors, `g_return_val_if_reached (-1);`
^^ though, of course, kernel really never should fail in such a manner (hence the assertion).

Also, is it nice to use g_get_monotonic_time()? GLib leaves it undefined that this is actually CLOCK_MONOTONIC (for good reasons), but libnm -- as an implementation -- should explicitly fallback to CLOCK_MONOTONIC -- despite itself leaving the actually used clock undefined (from a user's point of view)

Reply:
Okay. I'm not too comfortable about returning -1 -- the function should not be expected to ever fail, since even GLib itself can't work without CLOCK_MONOTONIC. But I don't really mind -- it's never going to be reached anyway.

Reply:
You are right. glib even crashes with `g_error()`, if it cannot get `CLOCK_MONOTONIC`. It's good to have functions that cannot fail -- especially, if it's not clear how to handle the error.